engine/class_modules/priest/sc_priest_shadow.cpp

```diff
@@ -885,18 +906,98 @@ struct vampiric_touch_t final : public priest_spell_t
 // ==========================================================================
 // Devouring Plague
 // ==========================================================================
+struct devouring_plague_dot_t final : public priest_spell_t
+{
+  double rolling_dp_bonus;
+
+  devouring_plague_dot_t( priest_t& p )
+    : priest_spell_t( "devouring_plague", p, p.find_class_spell( "Devouring Plague" ) ), rolling_dp_bonus( 0.0 )
+  {
+    may_crit                   = true;
+    tick_zero                  = false;
+    tick_may_crit              = true;
+    affected_by_shadow_weaving = true;
+    background                 = true;
+    hasted_ticks               = true;
+
+    base_dd_min = base_dd_max = spell_power_mod.direct = base_td = 0.0;
+  }
+
+  void init() override
+  {
+    priest_spell_t::init();
+
+    snapshot_flags |= STATE_CRIT | STATE_TGT_CRIT | STATE_HASTE;
+    update_flags |= STATE_CRIT | STATE_TGT_CRIT | STATE_HASTE;
+  }
+
+  // Make sure the dot doesn't also consume insanity
+  void consume_resource() override
+  {
+    return;
+  }
+
+  // TODO: see if we need this
+  // timespan_t composite_dot_duration( const action_state_t* s ) const override
+  // {
+  //   return dot_duration * ( tick_time( s ) / base_tick_time );
+  // }
+
+  void append_damage( player_t* target )
+  {
+    dot_t* dot = get_dot( target );
+    // get remaining full ticks left (not partials, these will be rolled over)
+    // calculate total damage
+    timespan_t remaining_dot = dot->remains();                                       // 2.5s
+    remaining_dot            = remaining_dot - dot->time_to_next_tick();             // 2.5 - .5s = 2s
+    int num_full_ticks = as<int>( std::ceil( remaining_dot / dot->time_to_tick ) );  // 2s / ~1.5 = 1.3 = 1 full tick
+
+    double total_damage = priest().tick_damage_over_time( num_full_ticks * dot->time_to_tick, dot );
```

Comment:
Why are you not just passing remaining_dot to tick_damage_over_time function here? Does the num_full_ticks calculation really make a difference?

And if it does: what happens when between the last tick of the dot where time_to_tick was calculated and now you gain/loose haste? Does it still take the last calculated time_to_tick or do you need to re-evaluate that with current haste?

Reply:
We're not passing the remaining time because that's not how it is implemented in-game. Currently, only remaining full ticks become increase in power for the next DP. Time_to_next_tick and the leftover partial tick get added as increased time to the new Devouring Plague. 

I.e. you cast a second Devouring Plague and there is 1 second to next tick, a full 1.5s tick, and a 0.4s long partial tick left on the first one. What happens is that full tick increases the damage of the ticks of the new Devouring Plague, distributed evenly between the next ticks. The 1s to next tick and the 0.4s partial tick get added as extra duration to the new Devouring Plague.

Haste in this mess is... complicated, we haven't observed it enough (because the in-game implementation has some bugs to say the least) to know exactly what haste entails here, but I believe it will just change future tick frequency without changing the new dot duration.

Reply:
Ok. I also somehow simplified the calculation in my commit, since there already is a dot state from which to calculate the tick_amount.